### PR TITLE
New SFDX post

### PR DESCRIPTION
**Pull request by marcantoineveilleux**

### CODE REVIEW

File: _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+author:
+  name: Marc-Antoine Veilleux
+  bio: Coveo for Salesforce Team Lead
+  image: maveilleux.jpg
+---
+
+
+I've been working with the Salesforce platform at Coveo for about 3 years. In those 3 years, I always had the feeling that there was something missing in our continuous integration setup. A couple of years ago, we've automated the process of creating a managed package using PhantomJS. As an ISV, this is a well-deserved upgrade.
+
+Since then, we were able to gain a lot of speed in our release process. But still, there was something missing. There was a void in my developer life. Demoing, reviewing, and testing new features on our package was near impossible and needed a lot of effort. Since Salesforce organization can't be created easily, a developer had to create a new org, push all the code to that org from their machine, give credentials to everyone, etc. A lot of _works on my machine(TM)_ and _Jenkins is weird_ hapenned using that method.
```
Comment (sbelzile-Coveo) [inline _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md]: hapenned => happened

File: _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+<!-- more -->
+
+## Salesforce DX
+
+At last year's Dreamforce, Salesforce introduced Salesforce DX and, since Winter '18, the wait is over. Salesforce DX is now GA and as a bonus, it's free
```
Comment (sbelzile-Coveo) [inline _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md]: free**.**

File: _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+To do so, you need 3 simple things: a way to authenticate to your Dev Hub, a scratch org definition, and the SFDX CLI. 
+
+Here's an example:
+
+```
+sfdx force:auth:web:login -s
+sfdx force:org:create -f config/scratch-def.json
+```
+
+And here's the `scratch-def.json`
```
Comment (sbelzile-Coveo) [inline _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md]: :

File: _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+### Installation
+
+Now that we know all those things, how can we integrate all of this inside our continuous integration system?
+
+Let's start by making sure SFDX is correctly installed on each machine and works on the developer machines. There are multiple options here: docker images, manual steps, and even wishful thinking. After some research, I've found out that Salesforce host a manifest file that contains all the needed URL to download SFDX for each different OS. 
+
+Then, using our source of truth, the next step is figuring out how to make it work for everyone. On previous projects, I've used the [phantomjs-prebuilt](http://phantomjs.org/) open source project to successfully install PhantomJs.
+
+I really liked the idea that, whatever the context (developer, windows, OS X, linux, Jenkins, Travis), it will install the needed binary correctly in the _install_ phase of NPM. That's the reason I've created [sfdx-prebuilt](https://github.com/coveo/sfdx-prebuilt). It's an open source project that will install, if needed, SFDX and return the right path to it. To use it, run a simple npm command:
```
Comment (sbelzile-Coveo) [inline _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md]: windows => **W**indows
linux => **L**inux

File: _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+})
+```
+
+### CLI Authentication
+
+Now that we have a good solution to run SFDX commands easily, the next problem is authentication.
+
+There are two ways you can authenticate to the SFDX CLI using only the command line: the `force:auth:sfdxurl:store` and the `force:auth:jwt:grant` commands. To the naked eye, both authentication seems equivalent but, don't get fooled: they are not.
+
+When you create a scratch org using the `force:auth:sfdxurl:store` command, you will not be able to re-authenticate the scratch org later on. By using the JWT token, you can easily and painlessly re-authenticate by simply using the scratch org username and the JWT token. So, right now, take my word for it: use the JWT flow. There is [good documentation to help you create all the needed certificate](https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_auth_jwt_flow.htm).
```
Comment (sbelzile-Coveo) [inline _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md]: re-authenticate **to? / into?** the scratch org later on

File: _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+When you have the certificate, I've personally decided to use environment variable to pass them along to the code. CI systems can easily inject environment variable and developers can set them easily on their machihe or use something like [dotenv](https://github.com/motdotla/dotenv).
+
+### Reuse Scratch Orgs
+
+One problem is still present: the fact that we must cleaning up unused scratch org.
```
Comment (sbelzile-Coveo) [inline _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md]: that we must cleaning up??? => that we must clean up

File: _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+There are some limits in Salesforce that force you to be smart about how many active scratch orgs you have. I've decided to go with one scratch org per SCM branch.
+
+The next problem is that, if you create any scratch org using the CI, the CI doesn't remember everything between each build. I've tried multiple things, but at the end of the day, I've decided to use the field called `AlmReference` on the `ActiveScratchOrg` object in Dev Hub.
```
Comment (sbelzile-Coveo) [inline _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md]: the CI, the CI => the CI, it?

File: _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+Basically, Salesforce has a default field specifically made so your build systems can add information. I've decided to feed the field with a custom JSON file containing the current branch name. Since scratch org are simply a record in the dev org, you can delete them easily using standard record manipulation API.
```
Comment (sbelzile-Coveo) [inline _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md]: using standard record manipulation API => using standard API record manipulation?

File: _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md [hunk 9]
```diff
@@ -11,15 +11,25 @@ author:
...
+I've been working with the Salesforce platform at Coveo for about 3 years. In those 3 years, I always had the feeling that there was something missing in our continuous integration setup. A couple of years ago, we've automated the process of creating a managed package using PhantomJS. As an ISV, this is a well-deserved upgrade.
+
+Since then, we were able to gain a lot of speed in our release process. But still, there was something missing. There was a void in my developer life. Demoing, reviewing, and testing new features on our package was near impossible and needed a lot of effort. Since Salesforce organization can't be created easily, a developer had to create a new org, push all the code to that org from their machine, give credentials to everyone, etc. A lot of _works on my machine(TM)_ and _Jenkins is weird_ hapenned using that method.
```
Comment (sbelzile-Coveo) [inline _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md]: Since Salesforce organization => Since **a** [...] ou organization**s**

File: _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md (hunk 8 above)
Comment (marcantoineveilleux) [reply _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md]: bof, je crois que les 2 sont valide?

Comment (sbelzile-Coveo) [reply _posts/2017-11-05-The-perfect-Salesforce-CI-setup-using-SFDX-NodeJS-and-NPM.md]: ça sonne drôle ta phrase. Si tu la gardes comme ça, met : "using **the** standard"

**Pull request merged**